### DISCUSSION TITLE
Docking support. I'm willing to try to implement something.

### DISCUSSION THREAD

**Discussion by DrOptix** [emilk/egui#424] [Ideas]
Before I create an issue and or a PR I would like to get some input about something.

Do you think egui would benefit from supporting docking panels?
I'm looking at ImGui docking and it looks awesome + it looks like you can actually detach a panel and it will use it's own native window to draw until you dock it again in the main window.

One draw back at the moment I can think for is that I don't know if we can support such a system on web, but should be relatively easy to put it behind a feature or platform gate.

I would like such a system from a game engine I started developing in my spare time. I'm willing to try and give it a shot especially that I need it for myself at some point.

@emilk, your opinion matters the most, so what do you think? Is there something like this in the works?

**Comment by DrOptix** [comment]
To be honest I don't really know how to best approach this one, especially the multiple native windows part when undocking.

At the moment I think I will continue my project without the idea of undocking the side panels or the main editor tabs. It would be great to have that option at some point, but the MVP for my project does not require docking.

If I come up with something I will let you know.

**Reply by Bigous** [reply]
You are mixing Docking and Viewport. With viewport you can drag the window outside the main window, creating another window when it happens and moving it back whe you drag it back. Docking can be done without the viewport (each window is dockable and have 5 reagions wher you can drag other windows into (top, left, bottom, right and center... and when it's in center you create tabs with each window).

**Comment by gbrls** [comment]
Hey, I just created [an issue](https://github.com/emilk/egui/issues/655) about this feature. I think if egui is trying to be Rust's imgui this feature is important!

**Comment by emilk** [comment]
Docking would be very nice, but it requires quite a bit of design work to think through how to implement it properly. We have panels now that can section up a window, so that's a start.

Perhaps it is possible to build this as a library initially. The user could define multiple dockable sections, each with their own id, name and closure. The docking library would then have the memory of whether or not each section is a separate window, a tab, or a panel inside a window.

Multiple native windows: let's punt on that, and keep it a bit simpler to start with.

In any case, I don't have the time/desire to work on this at the moment, but if someone wants to experiment then I can probably come with some guidance.

**Comment by namebee** [comment]
Have there been any further thoughts about this? It's the only thing preventing me from using egui over imgui-rs.

**Comment by JSH32** [comment]
Egui would be much better for editor type applications if it had docking support. Viewports (multiple windows) could be not included, but docking panels is extremely useful.

**Comment by wjbuys** [comment]
Looks like https://github.com/lain-dono/egui_docking exists, no docs but the example is cool

**Reply by emilk** [reply]
Hopefully @Adanos020 is working on it (according to https://github.com/lain-dono/egui_docking/issues/1)

**Reply by anhosh** [reply]
I am indeed! https://github.com/Adanos020/egui_dock

I discovered a few bugs along the way but I'm going to fix them after the 0.1.0 release. Right now I just want it to be accessible at its current state.

**Reply by anhosh** [reply]
Update: version 0.1.0 is now available on crates.io: https://crates.io/crates/egui_dock

I tried my best to provide the most complete documentation I could, but if there's anything you consider worth changing, feel free to contribute.

**Reply by yorqat** [reply]
@Adanos020 I recently looked into the thread and you are legendary
